### DISCUSSION TITLE
Terraform Stacks

### DISCUSSION THREAD

**Discussion by bakavets** [gruntwork-io/terragrunt#3510] [General]
Hi there! 

What do you think about Terraform Stacks (https://www.hashicorp.com/blog/terraform-stacks-explained)?

**Comment by josh-padnick** [comment]
We actually just announced Terragrunt Stacks in the [Road to Terragrunt 1.0 Blog Post](https://blog.gruntwork.io/the-road-to-terragrunt-1-0-4d5a0b416086)!

To offer a more nuanced response, for several years now, Terragrunt has operated on the premise that working in small, deployable units of infrastructure is the best way to scale with IaC. In fact, to copy from the very blog post you linked, this gives you:

- Limited blast radius
- Faster run times
- The ability to assign different permissions per unit
- The ability to deploy complex infrastructure in stages

But, again just as the blog post says, small, deployable units are not enough. You also need to _combine_ these units into reusable patterns. Terraform Stacks attempted to solve both of these problems at once. As I wrote above, Terragrunt has solved the "small, deployable units" problem for years, but with Terragrunt Stacks, we now aim to solve the "combine units into reusable patterns" problem.

Since Terragrunt has been working with small units for years, we've built up years of edge case handling, multiplexed logs, error management, dependency management, shared configuration, and more. We think re-creating that level of stability is highly non-trivial, and yet that's the foundation for working in stacks.

With Terragrunt Stacks, we're starting off by helping you leverage the `terragrunt.hcl` files you're already using. As we mature the feature, we'll focus more on "net new" use cases where you're defining a stack from scratch.

I hope that helps shed some light, and as always, we'd be grateful to hear any feedback you have on the above!

**Comment by josh-padnick** [comment]
Oh, I'd like to share one other thought since that word "stacks" has become so overloaded in our industry. We define two key terms:

- **Unit:** A single instance of an OpenTofu/Terraform module, represented as a `terragrunt.hcl`
- **Stack:** A collection of units (with optional configuration of those units), represented as a `terragrunt.stack.hcl`.

If it's helpful to compare with the Terraform ecosystem, whereas Terraform works in "workspaces" and now "stacks," Terragrunt works in "units" and also now "stacks."

**Comment by bakavets** [comment]
Hi @josh-padnick! Thank you for the reply!
